storage-proofs/porep/src/stacked/vanilla/cache.rs

```diff
@@ -134,21 +136,93 @@ impl ParentCache {
     {
         let path = cache_path(cache_entries, graph);
         if path.exists() {
-            Self::open(len, cache_entries, path)
+            Self::open(len, cache_entries, graph, path)
         } else {
             Self::generate(len, cache_entries, graph, path)
         }
     }
 
-    /// Opens an existing cache from disk.
-    pub fn open(len: u32, cache_entries: u32, path: PathBuf) -> Result<Self> {
-        info!("parent cache: opening {}", path.display());
+    /// Opens an existing cache from disk.  If the verify_cache option
+    /// is enabled, we rehash the data and compare with the persisted
+    /// hash file.  If the persisted hash file does not exist, we
+    /// re-generate the cache file, which will create it.
+    pub fn open<H, G>(
+        len: u32,
+        cache_entries: u32,
+        graph: &StackedGraph<H, G>,
+        path: PathBuf,
+    ) -> Result<Self>
+    where
+        H: Hasher,
+        G: Graph<H> + ParameterSetMetadata + Send + Sync,
+    {
+        let mut checksum_path = path.clone();
+        checksum_path.set_extension("cksum");
+
+        let mut verify_cache = settings::SETTINGS
+            .lock()
+            .expect("verify_cache settings lock failure")
+            .verify_cache;
+
+        info!(
+            "parent cache: opening {}, verify enabled: {}",
+            path.display(),
+            verify_cache
+        );
 
-        let cache = CacheData::open(0, len, &path)?;
-        info!("parent cache: opened");
+        // If the checksum file does not exist, generate the cache
+        // file again along with the checksum file.
+        if !Path::new(&checksum_path).exists() {
+            info!(
+                "[!!!] Parent cache checksum is missing.  Regenerating {}",
+                path.display()
+            );
+            ensure!(
+                Self::generate(len, graph.size() as u32, graph, path.clone()).is_ok(),
+                "Failed to generate parent cache"
+            );
+
+            // If we've just generated the checksum file, do not
+            // re-verify, even if requested.
+            verify_cache = false;
+        }
+
+        if verify_cache {
+            // Always check all of the data for integrity checks, even
+            // if we're only opening a portion of it.
+            let mut hasher = Sha256::new();
+            info!("[open] parent cache: calculating consistency digest");
+            let file = File::open(&path)?;
+            let data = unsafe {
+                memmap::MmapOptions::new()
+                    .map(&file)
+                    .with_context(|| format!("could not mmap path={}", path.display()))?
+            };
+            hasher.update(&data);
+            drop(data);
```

Comment:
it is likely more efficient to use `std::io::ReadBuf` with a regular file open, than mmap for just hashing the data

Reply:
Out of curiosity, do you know of any fast rust hashers/crates made specifically for large data/files that might fit this use?  I was wondering if something like that may work out all around better, but figured I'd stick with what we already have support for in the first cut.

Reply:
Given the systems we run on sha2 is likely good enough, the other fast alternative I could suggest is https://github.com/BLAKE3-team/BLAKE3

Reply:
Ok thanks, I tried blake2s since it was already supported, but switched back to sha2 ;-)  I'll give the ReadBuf approach a go.

Reply:
Was hoping for more of a speed-up, but if we loop over full buffers, it's slightly faster.  I'll update shortly.

Reply:
Very odd, on second test it appears much slower.  Hrm ...

Reply:
Dig, could you take a look at the commented out attempt at doing this?  I tested several times and it appears much slower, so I'm probably doing something wrong.

Reply:
Am I right to conclude dig's suggestion didn't turn out to be faster? (It seems not to have been taken in the current version.)

Reply:
As added, it was far slower in my testing, so I reverted.  I suspect dig may know a faster method or way of applying it, so it can be improved over time as needed.